Add comma-separated --auto-accept support.

johnharris85 opened the pull request:
Addresses #24143, simplifying `--auto-accept` syntax for a better UX.

Add support for comma-separated --auto-accept value.

`$ docker swarm update --auto-accept worker,manager`.

Signed-off-by: John Harris john@johnharris.io

johnharris85 committed:
Add support for comma-separated --auto-accept syntax.
Signed-off-by: John Harris <john@johnharris.io>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "fix-swarm-update-auto-accept" git@github.com:johnharris85/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

dnephin commented:
If we're going to support `all` instead of `--auto-accept worker  --auto-accept manager` I think we should make a much larger change to the `AutoAcceptOption`. 

The logic of `Set` becomes a lot simpler if you can't set it more than once.

johnharris85 commented:
I'm implementing the `--auto-accept-add` and `--auto-accept-rm` flags, but wondering how auto-acceptance should be handled on `docker swarm init`. Obviously `add` still works there but should `rm` error? Or just silently do nothing (as it looks like default behavior for init is going to be auto-accept = false for both anyway)?

dnephin commented:
I guess on `init` it should actually just use the current behaviour

thaJeztah commented:
I see the design discussion continued in the issue; https://github.com/docker/docker/issues/24143#issuecomment-230650677, so it looks like this PR needs to be updated

/cc @aaronlehmann @aluzzardi ?

johnharris85 commented:
Ye @thaJeztah just waiting for the design stuff to be fleshed out over there before updating the PR.

aluzzardi commented on api/client/swarm/opts.go:
```diff
@@ -112,6 +112,15 @@ func (o *AutoAcceptOption) Set(value string) error {
 		o.values[value] = true
+	case "ALL":
+		if accept, ok := o.values[WORKER]; ok && accept {
+			return fmt.Errorf("value NONE is incompatible with %s", WORKER)
```
This should just say `value NONE is compatible with ALL`

aluzzardi commented on api/client/swarm/opts.go:
```diff
@@ -112,6 +112,15 @@ func (o *AutoAcceptOption) Set(value string) error {
+			return fmt.Errorf("value NONE is incompatible with %s", WORKER)
+		}
+		if accept, ok := o.values[MANAGER]; ok && accept {
+			return fmt.Errorf("value NONE is incompatible with %s", MANAGER)
```
Same

johnharris85 replied:
Thanks @aluzzardi, missed these but actually the logic needs to be tightened and can be made simpler now I think about it (after adding `ALL`). Just running tests on my changes and then will update the PR.

johnharris85 commented:
Added support for comma-separated values `--auto-accept worker,manager` as well as `all`. Also backwards compatible with the current way of setting these flags. Added tests to reflect the changes.

cpuguy83 commented:
I don't think `all` is a good idea.
Today we accept `manager`, `worker` and it's ok, but in the future we may have other types (`storage`?) where it makes less sense to support `all`

cpuguy83 commented:
meh, on second thought it's probably fine.

johnharris85 commented:
@cpuguy83 `all` is a UX feature for me. I think there is more refactoring to be done here (to make `all` future proof for example) but the current syntax `--auto-accept worker --auto-accept manager` is too long-hand IMO. The `--auto-accept manager,worker` syntax is not currently supported (this PR adds it). 

`all` makes even more sense when considering the potential for future options. If I'm doing some local experiments / development and I need to type `--auto-accept n` where _n_ is maybe 4 or 5 options. That's going to get old pretty quickly :p

thaJeztah commented:
> `all` makes even more sense when considering the potential for future options.

I'm not sure we should develop for *possible* future options. What if the new future option is something that should be an "opt-in"? If someone set `all` (because: lazy), we could have a potential security issue. Better to keep it explicit, and have people update their config to allow the new type than to allow it without them knowing.

johnharris85 commented:
Maybe just allowing a comma-separated syntax is better then, and removing all? Would give a better UX shorthand instead of multiple `--auto-accept` flags but also preserve the security aspect?

aaronlehmann commented:
@johnharris85: I think that's fine.

thaJeztah commented:
Thanks! We can always add `all` back in future if we think it's ok; removing it is more troublesome once it's there

johnharris85 commented:
`all` removed and refactored some of the `none` checking logic to make it simpler. Tests amended.

aaronlehmann commented on api/client/swarm/opts.go:
```diff
@@ -18,6 +18,8 @@ const (
 	// MANAGER constant for manager name
 	MANAGER = "MANAGER"
+	// NONE constant
+	NONE = "NONE"
```
Not strictly related to your change, but could you please change these constants to CamelCase? Identifiers should never be all-caps in Golang. I'm surprised golint doesn't complain about this.

https://golang.org/doc/effective_go.html#mixed-caps
https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

aaronlehmann replied:
Also, they should not be exported. And then we can drop the silly comments.

johnharris85 replied:
They're being used in `opts_test.go`. So either they stay exported (but CamelCase) or we hard code the values into all the test cases?

aaronlehmann replied:
`opts_test.go` is in the same package, so they don't need to be exported.

johnharris85 replied:
Right, thanks.

aaronlehmann commented:
Can you please update the PR summary and description to reflect the changes from the review, and squash the commits?

aaronlehmann commented on api/client/swarm/opts.go:
```diff
@@ -99,27 +101,31 @@ func (o *AutoAcceptOption) String() string {
+		switch value {
+		case "", NONE:
+			// NONE must stand alone, so if any non-NONE setting exists, conflict
+			if !o.values[NONE] && len(o.values) > 0 {
```
I'd suggest moving this check outside of the loop, with something like:

```go
if o.values[NONE] && len(o.values) > 1 {
        ...
}
```

Among other things, this will give a more consistent error message that won't depend on the order of the values.

johnharris85 replied:
I actually don't see the value in specifying the explicit conflict, can't any error (where there is none + _n_) return with something like "value NONE cannot be specified alongside other node types" ?

johnharris85 replied:
Move to above line#105? `Set` is only called once though (and o.values is always empty until we set something in it) so that check will never eval to true right?

aaronlehmann replied:
Agreed, I don't see the value in specifying the conflict.

I was thinking this could move below the loop, so it inspects `o.values` after the map is populated.

johnharris85 replied:
Ah OK I get it. Ye can do, but with the standardizing of the error now does this have any other main benefits? I guess you could argue that error is checking the flag set as a whole, or that it's part of the `none` case. No strong feeling either way really.

aaronlehmann replied:
I think it makes most sense for it to be below the loop because otherwise you have to check for conflicts both in the `none` case (in case `none` comes after a different value) and in the other cases (in case they come after `none`). If the check is below the loop, only a single check is needed.

aaronlehmann commented on api/client/swarm/opts.go:
```diff
@@ -99,27 +101,31 @@ func (o *AutoAcceptOption) String() string {
+				return fmt.Errorf("value NONE is incompatible with %s", strings.Join(s, ", "))
+			}
+			o.values[WORKER] = false
+			o.values[MANAGER] = false
```
This seems wrong because elsewhere we're looking at `len(o.values)`. Setting a value to false won't change the number of values in the map. We should be deleting the value instead.

The real problem is that the type of the map is `map[string]bool`. It should be `map[string]struct{}` instead, because this map is being used as a set. Making that change will prevent this kind of error.

aaronlehmann replied:
Actually there shouldn't be a need to remove these items anyway, because this is unreachable if either one was set to true. Also, removing them would interfere with moving the compatibility check out of the loop, as I suggested in https://github.com/docker/docker/pull/24146/files#r70158372.

I would still highly recommend making the type of `o.values` `map[string]struct{}`, though.

johnharris85 replied:
Thanks for the feedback Aaron. Can you elaborate a little more on why this change would be more beneficial? (it was `map[string]bool` before I got to it, so just want to understand the benefits). Thanks!

aaronlehmann replied:
`map[string]struct{}` is a map that has keys but no value (think "map of empty struct"). So it's a useful way to implement a set data structure. If you use `map[string]bool`, you have to think about two cases: either the key isn't present, or it is present with a value of `false`. If you use `map[string]struct{}`, there's no value, so the only consideration is which keys are present and which aren't. I find it a little less error-prone, and it also signals to experienced Go coders that the map is being used as a set.

There is also a slight efficiency improvement from not storing any values, but it wouldn't matter for something like this.

aaronlehmann commented on api/client/swarm/opts.go:
```diff
@@ -86,40 +88,35 @@ func NewListenAddrOption() NodeAddrOption {
-	for key, value := range o.values {
-		keys = append(keys, fmt.Sprintf("%s=%v", strings.ToLower(key), value))
+	for key := range o.values {
+		keys = append(keys, fmt.Sprintf("%s=%v", strings.ToLower(key), true))
```
Just noticed this: should this be a list of values, instead of `x=true, y=true`? There is no `=false`, so the `=true` seems redundant.

johnharris85 replied:
Ye I thought that too, maybe just for readability / clarity (for repr), although ye, not strictly required as it's true if it exists by default. Happy to change but think it's clearer to stay. No strong preference either way though.

aaronlehmann replied:
Okay, if you decide to keep the current format, can you change the Sprintf call to `fmt.Sprintf("%s=true", strings.ToLower(key))`?

aaronlehmann commented on api/client/swarm/opts.go:
```diff
@@ -133,17 +130,32 @@ func (o *AutoAcceptOption) Policies(secret *string) []swarm.Policy {
+}
+
+// AddValue adds a value to the set of auto-accepted nodes
+func (o *AutoAcceptOption) AddValue(value string) {
```
Please unexport `IsPresent` and `AddValue` since they aren't meant to be used outside the package.

You might also consider removing `AddValue` and doing the map key assignment directly.

aaronlehmann commented on api/client/swarm/opts.go:
```diff
@@ -30,9 +30,11 @@ const (
+		{Role: manager, Autoaccept: false},
 	}
+
+	exists = struct{}{}
```
You can just use `struct{}{}` directly instead of defining a variable for it. It doesn't matter either way, but that's more typical usage.

LK4D4 replied:
I wonder what the point of using `struct{}` instead of bool. `bool` syntax is much simpler.

aaronlehmann replied:
@lk4d4 Check out the comments on outdated parts of the diff in this PR.

> On Jul 11, 2016, at 09:44, Alexander Morozov <notifications@github.com> wrote:
> 
> In api/client/swarm/opts.go:
> 
> >  	}
> > +
> > +	exists = struct{}{}
> I wonder what the point of using struct{} instead of bool. bool syntax is much simpler.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

LK4D4 replied:
@aaronlehmann heh, okay. I still think that readability of bool version is much better and I wouldn't prefer `struct{}` version for one-time set usage.

aaronlehmann commented on api/client/swarm/opts.go:
```diff
@@ -86,40 +86,35 @@ func NewListenAddrOption() NodeAddrOption {
+	for _, value := range strings.Split(acceptValues, ",") {
+		value = strings.ToUpper(value)
+		switch value {
+		case "", none:
```
Sorry, one final comment. I'm wondering it makes sense to treat `""` the same as `none`. It would mean that `worker,,manager` is parsed as `worker,none,manager`. I guess the intent here is to treat `--auto-accept ""` as `--auto-accept none`. If that's the case, maybe it would be better to do it at the very beginning of this function:

```go
if acceptValues == "" {
        o.values[none] = struct{}{}
} else {
        for _, value := range strings.Split(acceptValues, ",") {
                ...
        }
}
```

johnharris85 replied:
Good point, but then `worker,,manager` would trigger the `default` case? In addition to the above change, should a case be added for an empty string inside the loop to just ignore it?

aaronlehmann replied:
I'm not sure, but either sounds reasonable to me. Triggering an error in that case (by hitting the default case) is probably the safest approach.

johnharris85 replied:
And ye, I guess that was the original intent (that code existed before my PR) but I'm not sure whether I agree with it. I think a blank `--auto-accept` should error or pass silently, not set to `none`. That doesn't seem intuitive behaviour. Perhaps it was done for security reasons? In case someone set it accidentally but didn't pass anything, default to `none`. That may have made sense before, but aren't the defaults changing to `false` for both manager and worker anyway?

aaronlehmann replied:
> And ye, I guess that was the original intent (that code existed before my PR) but I'm not sure whether I agree with it. I think a blank --auto-accept should error or pass silently, not set to none.

Makes sense.

> That doesn't seem intuitive behaviour. Perhaps it was done for security reasons? In case someone set it accidentally but didn't pass anything, default to none.

Not really sure that makes sense. It's hard to accidentally pass an empty string as an argument.

> That may have made sense before, but aren't the defaults changing to false for both manager and worker anyway?

The autoaccept defaults aren't changing. What changed recently on master is that by default, a secret token is needed to join a swarm.

aaronlehmann commented:
Code LGTM

Can you please update the commit message and pull request description to summarize the final state of the changes?

This will also need updates to the docs before merging.

Thanks!

cpuguy83 commented:
LGTM

cpuguy83 commented:
ping @thaJeztah

thaJeztah commented:
Thanks! docs LGTM

aaronlehmann commented on docs/reference/commandline/swarm_init.md:
````diff
@@ -66,6 +66,13 @@ For example, the following initializes a cluster with auto-acceptance of workers
 ```
 
+It is possible to pass a comma-separated list of node types. The following initializes a cluster
+with auto-acceptance of both `worker` and `manager` nodes
````
I think `none` needs to be mentioned somewhere in the docs.

thaJeztah replied:
@aaronlehmann good point; just created https://github.com/docker/docker/pull/24552

aluzzardi commented:
Ping @icecrime
